--- uboot-custom-orig/arch/arm/dts/Makefile	2019-06-03 05:21:34.000000000 +0800
+++ uboot-custom-silicom/arch/arm/dts/Makefile	2020-06-15 11:14:20.933911632 +0800
@@ -87,7 +87,8 @@
 	tegra210-p2571.dtb
 
 dtb-$(CONFIG_ARCH_MVEBU) +=			\
-	armada-3720-db.dtb			\
+	ubmc_m.dtb
+#	armada-3720-db.dtb			\
 	armada-3720-ddr3-db-v2-B.dtb		\
 	armada-3720-ddr3-db-v2-C.dtb		\
 	armada-3720-ddr3-db-v1-A.dtb		\
--- uboot-custom-orig/arch/arm/dts/ubmc_m.dts	1970-01-01 08:00:00.000000000 +0800
+++ uboot-custom-silicom/arch/arm/dts/ubmc_m.dts	2020-06-17 12:27:36.000000000 +0800
@@ -0,0 +1,537 @@
+/*
+ * Device Tree file for Globalscale Marvell ESPRESSOBin Board
+ * Copyright (C) 2016 Marvell
+ *
+ * Romain Perier <romain.perier@free-electrons.com>
+ *
+ * This file is dual-licensed: you can use it either under the terms
+ * of the GPL or the X11 license, at your option. Note that this dual
+ * licensing only applies to this file, and not this project as a
+ * whole.
+ *
+ *  a) This file is free software; you can redistribute it and/or
+ *     modify it under the terms of the GNU General Public License as
+ *     published by the Free Software Foundation; either version 2 of the
+ *     License, or (at your option) any later version.
+ *
+ *     This file is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU General Public License for more details.
+ *
+ * Or, alternatively,
+ *
+ *  b) Permission is hereby granted, free of charge, to any person
+ *     obtaining a copy of this software and associated documentation
+ *     files (the "Software"), to deal in the Software without
+ *     restriction, including without limitation the rights to use,
+ *     copy, modify, merge, publish, distribute, sublicense, and/or
+ *     sell copies of the Software, and to permit persons to whom the
+ *     Software is furnished to do so, subject to the following
+ *     conditions:
+ *
+ *     The above copyright notice and this permission notice shall be
+ *     included in all copies or substantial portions of the Software.
+ *
+ *     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ *     EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ *     OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ *     NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ *     HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ *     WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ *     FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ *     OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/dts-v1/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include "ubmc_m.dtsi"
+
+/ {
+	model = "Silicom uBMC";
+	compatible = "globalscale,espressobin", "marvell,armada3720", "marvell,armada3710";
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+		/*stdout-path = "serial1:115200n8";*/
+	};
+	aliases {
+		ethernet0 = &eth0;
+		/*i2c0 = &i2c0;*/
+		/*i2c1 = &i2c1;*/
+		spi0 = &spi0;
+	};
+	memory@0 {
+		device_type = "memory";
+		reg = <0x00000000 0x00000000 0x00000000 0x40000000>;
+
+	};
+	vcc_sd_reg0: regulator@0 {
+		compatible = "regulator-gpio";
+		regulator-name = "vcc_sd0";
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <3300000>;
+		regulator-type = "voltage";
+		states = <1800000 0x1
+			  3300000 0x0>;
+		gpios = <&gpionb 4 GPIO_ACTIVE_HIGH>;
+	};
+	/*
+	vcc_sd_reg1: regulator {
+		compatible = "regulator-gpio";
+		regulator-name = "vcc_sd1";
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <3300000>;
+		regulator-boot-on;
+
+		gpios = <&gpionb 4 GPIO_ACTIVE_HIGH>;
+		gpios-states = <0>;
+		states = <1800000 0x1
+			  3300000 0x0>;
+		enable-active-high;
+	};
+	*/
+	
+	mdio@32004 {
+		#address-cells = <1>;
+		#size-cells = <0>;
+		status = "okay";
+
+		mvswitch: switch@000000{
+			compatible = "marvell,mv88e6085","marvell,mv88e6xxx","marvell,mv88e6341";
+			status = "okay";
+			phy-addr = <0>;
+			cpu-port = <0x0>;
+			/*port-mask = <0xf>;*/
+			port-mask = <0x15>;
+			/*port-mask = <0x3f>;*/
+			ports {
+			#address-cells = <1>;
+			#size-cells = <0>;
+/*
+			port@0 {
+				reg = <0>;
+				label = "cpu";
+				ethernet = <&eth0>;
+			};
+
+			port@1 {
+				reg = <1>;
+				label = "wan";
+				phy-handle = <&switch0phy0>;
+			};
+
+			port@2 {
+				reg = <2>;
+				label = "lan0";
+				phy-handle = <&switch0phy1>;
+			};
+
+			port@3 {
+				reg = <3>;
+				label = "lan1";
+				phy-handle = <&switch0phy2>;
+			};
+*/
+		};
+
+		mdio {
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			switch0phy0: switch0phy0@11 {
+				reg = <0x11>;
+			};
+			switch0phy1: switch0phy1@12 {
+				reg = <0x12>;
+			};
+			switch0phy2: switch0phy2@13 {
+				reg = <0x13>;
+			};
+		};
+		};
+	};
+	
+	reg_usb3_vbus: usb3_vbus@0 {
+		compatible = "regulator-fixed";
+		regulator-name = "usb3-vbus";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		shutdown-delay-us = <1000000>;
+		gpio = <&expander0 1 GPIO_ACTIVE_HIGH>;
+	};
+	reg_usb2_vbus: usb2-vbus {
+		compatible = "regulator-fixed";
+		regulator-name = "usb2-vbus";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		enable-active-high;
+		regulator-always-on;
+		gpio = <&expander0 0 GPIO_ACTIVE_HIGH>;
+	};
+	usb3_phy: usb3-phy {
+		compatible = "usb-nop-xceiv";
+		vcc-supply = <&reg_usb3_vbus>;
+	};
+	usb2_phy: usb2-phy {
+		compatible = "usb-nop-xceiv";
+		vcc-supply = <&reg_usb2_vbus>;
+	};
+	
+};
+
+&comphy {
+	max-lanes = <3>;
+	phy0 {
+			reg = <0>;
+			#phy-cells = <1>;
+			phy-type = <COMPHY_TYPE_USB3_DEVICE>;
+			phy-speed = <COMPHY_SPEED_5G>;
+	};
+
+	phy1 {
+			reg = <1>;
+			#phy-cells = <1>;
+			phy-type = <COMPHY_TYPE_PEX0>;
+			phy-speed = <COMPHY_SPEED_2_5G>;
+	};
+
+	phy2 {
+			reg = <2>;
+			#phy-cells = <1>;
+			phy-type = <COMPHY_TYPE_SATA0>;
+			phy-speed = <COMPHY_SPEED_6G>;
+	};
+};
+
+/* J9 */
+
+&pcie0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pcie_pins>;
+	reset-gpio = <&gpiosb 3 GPIO_ACTIVE_HIGH>;
+	status = "disabled";
+};
+
+
+/* J6 */
+&sata {
+	status = "okay";
+};
+/* J1 */
+/*sd card not used*/
+/*
+&sdhci1 {
+	wp-inverted;
+	bus-width = <4>;
+	cd-gpios = <&gpionb 3 GPIO_ACTIVE_LOW>;
+	marvell,pad-type = "sd";
+	vqmmc-supply = <&vcc_sd_reg1>;
+	status = "disabled";
+};
+*/
+/*emmc*/
+&sdhci1 {
+	non-removable;
+	bus-width = <4>;
+	/*mmc-hs200-1_8v;*/
+	mmc-ddr-1_8v;
+	/*max-frequency = <50000000>;*/
+	mmc-hs400-1_8v;
+	marvell,pad-type = "fixed-1-8v";
+	/*marvell,xenon-phy-slow-mode;*/
+	pinctrl-names = "default";
+	pinctrl-0 = <&mmc_pins>;
+	status = "okay";
+	#address-cells = <1>;
+	#size-cells = <0>;
+	mmccard: mmccard@0 {
+		compatible = "mmc-card";
+		reg = <0>;
+	};
+	
+};
+
+&i2c0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&i2c1_pins>;
+	status = "okay";
+	clock-frequency = <100000>;
+	expander0: pca953x@22 {
+		compatible = "nxp,pca9555";
+        #gpio-cells = <2>;
+		reg = <0x22>;
+		status = "disabled";
+	};
+};
+&i2c1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&i2c2_pins>;
+	status = "okay";
+	clock-frequency = <100000>;
+	bmc_rtc: pcf8523tk@68{
+		compatible = "nxp,pcf8523";
+		reg = <0x68>;
+		status = "okay";
+	};
+	bmc_fru: eeprom@50 {
+		compatible = "atmel,24c02";
+		reg = <0x50>;
+		pagesize = <16>;
+		size = <256>;
+	};
+	host_fru: eeprom@56 {
+		compatible = "atmel,24c02";
+		reg = <0x56>;
+	};
+	host_fan: max31785@53 {
+		compatible = "maxim,max31785";
+		reg = <0x53>;
+	};
+	host_adc: ads7830@48 {
+		compatible = "ti,ads7830";
+		reg = <0x48>;
+	};
+};
+
+/* Exported on the micro USB connector J5 through an FTDI */
+&uart0 {
+	status = "okay";
+};
+&uart1 {
+	status = "okay";
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart2_pins>;
+};
+/* J7 */
+#if 1
+&usb3 {
+	vbus-supply = <&reg_usb3_vbus>;
+	status = "disable";
+};
+#endif
+
+#if 0
+&usb3 {
+	status = "disabled";
+	usb-phy = <&usb3_phy>;
+	dr_mode = "device"
+	/* Generic PHY, providing serdes lanes */
+	phys = <&comphy 2>;
+	phy-names = "usb";
+};
+#endif
+
+&u3d{
+	status = "okay";
+};
+&usb2 {
+	status = "okay";
+	usb-phy = <&usb2_phy>;
+	phys = <&utmi_usb2>;
+	phy-names = "usb";
+};
+
+
+&utmi_usb32 {
+	status = "okay";
+};
+&utmi_usb2 {
+	status = "okay";
+};
+
+/*&eth0 {
+	phy-mode = "rgmii-id";
+	status = "okay";
+
+	fixed-link {
+		speed = <1000>;
+		full-duplex;
+	};
+};
+*/
+&eth0 {
+	status = "okay";
+	pinctrl-names = "default";
+	pinctrl-0 = <&rgmii_pins>, <&smi_pins>;
+	phy-mode = "rgmii-id";
+	phy_addr = <0x1>;
+	phy = <&mvswitch>;
+
+	fixed-link {
+		speed = <1000>;
+		full-duplex;
+	};
+};
+&eth1 {
+	status = "okay";
+	/* Generic PHY, providing serdes lanes */
+	/*phys = <&a3700_comphy0 1>;*/
+	phy-mode = "sgmii";
+	/*phy = <&phy1>;*/
+
+	fixed-link {
+		speed = <1000>;
+		full-duplex;
+	};
+
+};
+&spi0 {
+	status = "okay";
+	flash@0 {
+		reg = <0>;
+		compatible = "winbond,w25q32dw", "jedec,spi-flash";
+		spi-max-frequency = <10000000>;
+		m25p,fast-read;
+	};
+	w25q256@1 {
+		reg = <1>;
+		compatible =  "winbond,w25q256", "jedec,spi-nor";
+		spi-max-frequency = <30000000>;
+		m25p,fast-read;
+	};
+};
+/*GPIO1 */
+/*&pinctrl_nb GPIO1 gpiochip0 476*/
+&gpionb
+{
+	ngpios = <36>;
+	gpio-line-names =
+		"0",
+		"1",
+		"2",
+		"3",
+		"4",
+		"LED_BMC_BOT_RED_N",	/* "5"  481,*/
+		"LED_BMC_BOT_GRN_N",	/*"6",	482*/
+		"LED_BMC_BOT_BLU_N",	/*"7",	483*/
+		"8",
+		"9",
+		"10",
+		"HOST_PLTRST_N",		/*"11",487 */
+		"HOST_S45_N",			/*"12",	488*/
+		"HOST_S3_N",			/*"13",489*/
+		"HOST_ERROR_N",			/*"14",490*/
+		"HOST_THERMTRIP_N",		/*"15",491*/
+		"SPI_HOST_MUX_EN",		/*"16",492*/
+		"SPI_BMC_CS1",
+		"18",
+		"19",
+		"20",
+		"21",
+		"22",
+		"23",
+		"24",
+		"25",
+		"26",
+		"27",
+		"28",
+		"29",
+		"30",
+		"31",
+		"32",
+		"33",
+		"34",
+		"35";
+};
+/*&pinctrl_sb GPIO2 gpiochip1 446*/
+&gpiosb
+{
+	ngpios = <30>;
+	gpio-line-names =
+		"HOST_PROCHOT_N",		/*"0",*/
+		"1",
+		"RS232_CABLE_DETECTED",		/*"2",448*/
+		"DYING_GASP_N",			/*"3",449*/
+		"GPIO_HOST_1",			/*"4" 450,*/
+		"BMC_UART_MUX_SEL",		/*"5" 451,*/
+		"6",
+		"7",
+		"8",
+		"9",
+		"10",
+		"11",
+		"12",
+		"13",
+		"14",
+		"15",
+		"16",
+		"17",
+		"18",
+		"19",
+		"20",
+		"BMC_FP_BUTTON_OVERRIDE",	/*"21",467*/
+		"FP_RSTBTN_N",				/*"22",468*/
+		"FP_PWRBTN_N",				/*"23",469*/
+		"HOST_BOOT_COMPLETE_N",		/*"24",470*/
+		"PSU_LEFT_PWRGD",			/*"25",471*/
+		"BMC_RSTBTN_OUT_N",			/*"26",472*/
+		"HOST_FRU_EEPROM_MUX_SEL",		/*"27",473*/
+		"PSU_RIGHT_PWRGD",			/*"28",474*/
+		"BMC_PWRBTN_OUT_N";			/*"29";475*/
+};
+#if 0
+/*for espressobin support*/
+&spi0 {
+	status = "okay";
+	pinctrl-names = "default";
+	pinctrl-0 = <&spi_quad_pins>;
+	clock-frequency = <20000000>;
+	spi-max-frequency = <20000000>;
+	flash@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		mode = <0>;
+		reg = <0>;
+		compatible = "winbond,w25q32dw", "jedec,spi-flash";
+		spi-max-frequency = <10000000>;
+		m25p,fast-read;
+		};
+};
+#endif
+#if 0
+/*
+&spi0 {
+	status = "okay";
+	pinctrl-names = "default";
+	pinctrl-0 = <&spi_quad_pins>;
+	/*clock-frequency = <200000000>;*/
+	spi-flash@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "st,m25p128", "spi-flash";
+		reg = <0>; /* Chip select 0 */
+		spi-max-frequency = <40000>;
+		/*spi-max-frequency = <133000000>;*/
+		m25p,fast-read;
+	};
+};
+*/
+#endif
+
+#if 0
+/*for uBMC spi flash*/
+&spi0 {
+	status = "okay";
+	pinctrl-names = "default";
+	pinctrl-0 = <&spi_quad_pins>;
+	/* sqf_clk is a /4 from 800MHz TBG_A_P clock, set here because default
+	 * in driver is 160MHz, which is wrong
+	 */
+	clock-frequency = <20000000>;
+	/* limit bus to 20Mhz since the level shifters add load */
+	spi-max-frequency = <20000000>;
+	spi-flash@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		mode = <0>;
+		/*compatible = "winbond,w25q32", "jedec,spi-nor", "spi-flash";*/
+		compatible = "w25q32", "jedec,spi-nor", "spi-flash";
+		/*compatible = "w25q32jv,spi-nor";*/
+		reg = <0>; /* Chip select 0 */
+		spi-max-frequency = <10000000>;
+		m25p,fast-read;
+	};
+};
+#endif
--- uboot-custom-orig/arch/arm/dts/ubmc_m.dtsi	1970-01-01 08:00:00.000000000 +0800
+++ uboot-custom-silicom/arch/arm/dts/ubmc_m.dtsi	2020-06-16 17:59:15.641088625 +0800
@@ -0,0 +1,586 @@
+/*
+ * Device Tree Include file for Marvell Armada 37xx family of SoCs.
+ *
+ * Copyright (C) 2016 Marvell
+ *
+ * Gregory CLEMENT <gregory.clement@free-electrons.com>
+ *
+ * This file is dual-licensed: you can use it either under the terms
+ * of the GPL or the X11 license, at your option. Note that this dual
+ * licensing only applies to this file, and not this project as a
+ * whole.
+ *
+ *  a) This file is free software; you can redistribute it and/or
+ *     modify it under the terms of the GNU General Public License as
+ *     published by the Free Software Foundation; either version 2 of the
+ *     License, or (at your option) any later version.
+ *
+ *     This file is distributed in the hope that it will be useful
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU General Public License for more details.
+ *
+ * Or, alternatively
+ *
+ *  b) Permission is hereby granted, free of charge, to any person
+ *     obtaining a copy of this software and associated documentation
+ *     files (the "Software"), to deal in the Software without
+ *     restriction, including without limitation the rights to use
+ *     copy, modify, merge, publish, distribute, sublicense, and/or
+ *     sell copies of the Software, and to permit persons to whom the
+ *     Software is furnished to do so, subject to the following
+ *     conditions:
+ *
+ *     The above copyright notice and this permission notice shall be
+ *     included in all copies or substantial portions of the Software.
+ *
+ *     THE SOFTWARE IS PROVIDED , WITHOUT WARRANTY OF ANY KIND
+ *     EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ *     OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ *     NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ *     HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY
+ *     WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ *     FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ *     OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <dt-bindings/interrupt-controller/arm-gic.h>
+#include <dt-bindings/comphy/comphy_data.h>
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/phy/phy-utmi-mvebu.h>
+/ {
+	model = "Marvell Armada 37xx SoC";
+	compatible = "marvell,armada3700";
+	interrupt-parent = <&gic>;
+	#address-cells = <2>;
+	#size-cells = <2>;
+
+	aliases {
+		serial0 = &uart0;
+		serial1 = &uart1;
+	};
+
+	cpus {
+		#address-cells = <1>;
+		#size-cells = <0>;
+		cpu@0 {
+			device_type = "cpu";
+			compatible = "arm,cortex-a53", "arm,armv8";
+			reg = <0>;
+			enable-method = "psci";
+		};
+	};
+
+	psci {
+		compatible = "arm,psci-0.2";
+		method = "smc";
+	};
+
+	timer {
+		compatible = "arm,armv8-timer";
+		interrupts = <GIC_PPI 13 IRQ_TYPE_LEVEL_HIGH>,
+			     <GIC_PPI 14 IRQ_TYPE_LEVEL_HIGH>,
+			     <GIC_PPI 11 IRQ_TYPE_LEVEL_HIGH>,
+			     <GIC_PPI 10 IRQ_TYPE_LEVEL_HIGH>;
+	};
+
+	pmu {
+		compatible = "arm,armv8-pmuv3";
+		interrupts = <GIC_PPI 7 IRQ_TYPE_LEVEL_HIGH>;
+	};
+
+	soc {
+		compatible = "simple-bus";
+		#address-cells = <2>;
+		#size-cells = <2>;
+		ranges;
+
+		internal-regs@d0000000 {
+			#address-cells = <1>;
+			#size-cells = <1>;
+			compatible = "simple-bus";
+			/* 32M internal register @ 0xd000_0000 */
+			ranges = <0x0 0x0 0xd0000000 0x2000000>;
+
+			spi0: spi@10600 {
+				compatible = "marvell,armada-3700-spi";
+				#address-cells = <1>;
+				#size-cells = <0>;
+				reg = <0x10600 0xA00>;
+				clocks = <&nb_periph_clk 7>;
+				interrupts = <GIC_SPI 0 IRQ_TYPE_LEVEL_HIGH>;
+				num-cs = <4>;
+				status = "disabled";
+			};
+#if 0		
+			comphy: comphy@18300 {
+				compatible = "marvell,mvebu-comphy", "marvell,comphy-armada-3700";
+				reg = <0x18300 0x28>,
+				      <0x1f300 0x3d000>;
+				mux-bitcount = <4>;
+				max-lanes = <3>;
+			};
+#endif
+			
+#if 1
+			/*Modifid as kernel dtsi*/
+			comphy: comphy@18300 {
+				compatible = "marvell,mvebu-comphy", "marvell,comphy-a3700", "marvell,comphy-armada-3700";
+				reg = <0x18300 0x28>,
+				      <0x1f300 0x3d000>,
+					  <0x5c000 0x1000>, <0xe0178 0x8>;
+				reg-names = "comphy", "serdes",
+					    "usb3_gbe1_phy", "indirect_ahci";
+				#address-cells = <1>;
+				#size-cells = <0>;
+				mux-bitcount = <4>;
+				/*max-lanes = <3>;*/
+#if 0
+				a3700_comphy0: phy@0 {
+					reg = <0>;
+					#phy-cells = <1>;
+				};
+
+				a3700_comphy1: phy@1 {
+					reg = <1>;
+					#phy-cells = <1>;
+				};
+
+				a3700_comphy2: phy@2 {
+					reg = <2>;
+					#phy-cells = <1>;
+				};
+#endif
+			};
+#endif
+			i2c0: i2c@11000 {
+				compatible = "marvell,armada-3700-i2c";
+				reg = <0x11000 0x24>;
+				#address-cells = <1>;
+				#size-cells = <0>;
+				clocks = <&nb_periph_clk 10>;
+				interrupts = <GIC_SPI 1 IRQ_TYPE_LEVEL_HIGH>;
+			/*	mrvl,i2c-fast-mode;*/
+				status = "disabled";
+			};
+
+			i2c1: i2c@11080 {
+				compatible = "marvell,armada-3700-i2c";
+				reg = <0x11080 0x24>;
+				#address-cells = <1>;
+				#size-cells = <0>;
+				clocks = <&nb_periph_clk 9>;
+				interrupts = <GIC_SPI 2 IRQ_TYPE_LEVEL_HIGH>;
+			/*	mrvl,i2c-fast-mode;*/
+				status = "disabled";
+			};
+
+			uart0: serial@12000 {
+				compatible = "marvell,armada-3700-uart";
+				reg = <0x12000 0x200>;
+				interrupts = <GIC_SPI 11 IRQ_TYPE_LEVEL_HIGH>;
+				status = "disabled";
+			};
+			uart1: serial@12200 {
+				compatible = "marvell,armada-3700-uart-ext";
+				reg = <0x12200 0x30>;
+				clocks = <&xtalclk>;
+				interrupts =
+				<GIC_SPI 30 IRQ_TYPE_EDGE_RISING>,
+				<GIC_SPI 31 IRQ_TYPE_EDGE_RISING>;
+				interrupt-names = "uart-tx", "uart-rx";
+				status = "disabled";
+			};
+			nb_periph_clk: nb-periph-clk@13000 {
+				compatible = "marvell,armada-3700-periph-clock-nb";
+				reg = <0x13000 0x100>;
+				clocks = <&tbg 0>, <&tbg 1>, <&tbg 2>,
+				<&tbg 3>, <&xtalclk>;
+				#clock-cells = <1>;
+			};
+
+			sb_periph_clk: sb-periph-clk@18000 {
+				compatible = "marvell,armada-3700-periph-clock-sb";
+				reg = <0x18000 0x100>;
+				clocks = <&tbg 0>, <&tbg 1>, <&tbg 2>,
+				<&tbg 3>, <&xtalclk>;
+				#clock-cells = <1>;
+			};
+
+			tbg: tbg@13200 {
+				compatible = "marvell,armada-3700-tbg-clock";
+				reg = <0x13200 0x100>;
+				clocks = <&xtalclk>;
+				#clock-cells = <1>;
+			};
+
+			pinctrl_nb: pinctrl@13800 {
+				compatible = "marvell,armada3710-nb-pinctrl",
+					     "syscon", "simple-mfd";
+				reg = <0x13800 0x100>, <0x13C00 0x20>;
+				gpionb: gpio {
+					#gpio-cells = <2>;
+					gpio-ranges = <&pinctrl_nb 0 0 36>;
+					gpio-controller;
+					interrupts =
+					<GIC_SPI 51 IRQ_TYPE_LEVEL_HIGH>,
+					<GIC_SPI 52 IRQ_TYPE_LEVEL_HIGH>,
+					<GIC_SPI 53 IRQ_TYPE_LEVEL_HIGH>,
+					<GIC_SPI 54 IRQ_TYPE_LEVEL_HIGH>,
+					<GIC_SPI 55 IRQ_TYPE_LEVEL_HIGH>,
+					<GIC_SPI 56 IRQ_TYPE_LEVEL_HIGH>,
+					<GIC_SPI 57 IRQ_TYPE_LEVEL_HIGH>,
+					<GIC_SPI 58 IRQ_TYPE_LEVEL_HIGH>,
+					<GIC_SPI 152 IRQ_TYPE_LEVEL_HIGH>,
+					<GIC_SPI 153 IRQ_TYPE_LEVEL_HIGH>,
+					<GIC_SPI 154 IRQ_TYPE_LEVEL_HIGH>,
+					<GIC_SPI 155 IRQ_TYPE_LEVEL_HIGH>;
+
+				};
+
+				xtalclk: xtal-clk {
+					compatible = "marvell,armada-3700-xtal-clock";
+					clock-output-names = "xtal";
+					#clock-cells = <0>;
+				};
+				/*GPIO1*/
+				/*pin15-pin16*/
+				spi_quad_pins: spi-quad-pins {
+					groups = "spi_quad";
+					function = "gpio";
+				};
+				/*pin17*/
+				/*
+				spi_cs1_pins:spi-cs1-pins {
+					groups = "spi_cs1";
+					function = "spi";
+				};
+				*/
+				/*pin20-pin24*/
+				jtag_pins: jtag-pins {
+					groups = "jtag";
+					function = "jtag";
+				};
+				/*pin0-pin1*/
+				i2c1_pins: i2c1-pins {
+					groups = "i2c1";
+					function = "i2c";
+				};
+				/*pin2-pin3*/
+				i2c2_pins: i2c2-pins {
+					groups = "i2c2";
+					function = "i2c";
+				};
+				/*pin4*/
+				onewire_pins:onewire-pins{
+					groups = "onewire";
+					function = "gpio";
+				};
+				/*pin6*/
+				pmic0_pins:pmic0-pins{
+					groups = "pmic0";
+					function = "gpio";
+				};
+				/*pin8-pin10*/
+				sdio0_pins:sdio0-pins{
+					groups = "sdio0";
+					function = "gpio";	
+				};
+				/*pin7*/
+				pmic1_pins:pmic1-pins{
+					groups = "pmic1";
+					function = "gpio";
+				};
+				/*pin11*/
+				pwm0_pins:pwm0-pins{
+					groups = "pwm0";
+					function = "gpio";
+				};
+				/*pin12*/
+				pwm1_pins:pwm1-pins{
+					groups = "pwm1";
+					function = "gpio";
+				};
+				/*pin13*/
+				pwm2_pins:pwm2-pins{
+					groups = "pwm2";
+					function = "gpio";
+				};
+				/*pin14*/
+				pwm3_pins:pwm3-pins{
+					groups = "pwm3";
+					function = "gpio";
+				};
+				/*pin25-pin26*/
+				uart1_pins: uart1-pins {
+					groups = "uart1";
+					function = "uart";
+				};
+				/*pin9-pin10 pin18-pin19*/
+				uart2_pins: uart2-pins {
+					groups = "uart2";
+					function = "uart";
+				};
+				/*pin27-pin35*/
+				mmc_pins: mmc-pins {
+					groups = "emmc_nb";
+					function = "emmc";
+				};
+			};
+
+			pinctrl_sb: pinctrl@18800 {
+				compatible = "marvell,armada3710-sb-pinctrl",
+					     "syscon", "simple-mfd";
+				reg = <0x18800 0x100>, <0x18C00 0x20>;
+				gpiosb: gpio {
+					#gpio-cells = <2>;
+					gpio-ranges = <&pinctrl_sb 0 0 30>;
+					gpio-controller;
+					interrupts =
+					<GIC_SPI 160 IRQ_TYPE_LEVEL_HIGH>,
+					<GIC_SPI 159 IRQ_TYPE_LEVEL_HIGH>,
+					<GIC_SPI 158 IRQ_TYPE_LEVEL_HIGH>,
+					<GIC_SPI 157 IRQ_TYPE_LEVEL_HIGH>,
+					<GIC_SPI 156 IRQ_TYPE_LEVEL_HIGH>;
+				};
+				/*GPIO2*/
+				/*pin0*/
+				usb32_drvvbus0_pins: usb32-drvvbus0-pins
+				{
+					groups = "usb32_drvvbus0";
+					function = "drvbus";
+				};
+				/*pin1*/
+				usb2_drvvbus1_pins: usb2-drvvbus1-pins
+				{
+					groups = "usb2_drvvbus1";
+					function = "drvbus";	
+				};
+				/*pin24-pin29*/
+				sdio_pins: sdio-pins {
+					groups = "sdio_sb";
+					/*function = "sdio";*/
+					function = "gpio";
+				};
+				/*pin20-pin22*/
+				ptp_pins: ptp-pins {
+					groups = "ptp";
+					function = "gpio";
+				};
+				/*pin6-pin17*/
+				rgmii_pins: mii-pins {
+					groups = "rgmii";
+					function = "mii";
+				};
+				/*pin3-pin5*/
+				pcie_pins: pcie-pins {
+					groups = "pcie1";
+					function = "gpio";
+				};
+				/*pin18-pin19*/
+				smi_pins: smi-pins {
+					groups = "smi";
+					function = "smi";
+				};
+				/*pin23*/
+				mii_col_pins: mii-col-pins{
+					groups = "mii_col";
+					function = "gpio";
+				};
+
+			};
+
+			eth0: ethernet@30000 {
+				   compatible = "marvell,armada-3700-neta";
+				   reg = <0x30000 0x4000>;
+				   interrupts = <GIC_SPI 42 IRQ_TYPE_LEVEL_HIGH>;
+				   clocks = <&sb_periph_clk 8>;
+				   status = "disabled";
+			};
+
+			mdio: mdio@32004 {
+				#address-cells = <1>;
+				#size-cells = <0>;
+				compatible = "marvell,orion-mdio";
+				reg = <0x32004 0x4>;
+			};
+
+			eth1: ethernet@40000 {
+				compatible = "marvell,armada-3700-neta";
+				reg = <0x40000 0x4000>;
+				interrupts = <GIC_SPI 45 IRQ_TYPE_LEVEL_HIGH>;
+				clocks = <&sb_periph_clk 7>;
+				status = "disabled";
+			};
+			#if 0
+			u3d: u3d@50000 {
+				status = "disabled";
+				 /* NUM1 phys =  <&a3700_comphy 0 0x21fc>; */
+				/* NUM2 phys =  <&a3700_comphy 0 0x31fc>; */
+				phys = <&a3700_comphy0 0xa0fc>; /* NUM3 */
+				/* Either NUM1,NUM2,NUM3 corresponds to mode COMPHY_USB3 because
+				these are the only three modes and I have tested
+				phys = <&a3700_comphy 0 COMPHY_USB3> and the kernel does not complain
+                               */
+				phy-names = "usb";
+			};
+			#endif
+			/*see /Documentation/devicetree/bindings/phy/phy-mvebu.txt*/
+			u3d: u3d@50000 {
+				status = "disabled";
+				compatible = "marvell,armada3700-u3d";
+				 /* NUM1 phys =  <&a3700_comphy 0 0x21fc>; */
+				/* NUM2 phys =  <&a3700_comphy 0 0x31fc>; */
+				phys = <&utmi_usb32>; /* NUM3 */
+				phy-names = "usb";
+			};
+			usb3: usb@58000 {
+				compatible = "marvell,armada3700-xhci",
+				"generic-xhci";
+				reg = <0x58000 0x4000>;
+				interrupts = <GIC_SPI 15 IRQ_TYPE_LEVEL_HIGH>;
+				status = "disabled";
+			};
+
+			usb2: usb@5e000 {
+				compatible = "marvell,armada-3700-ehci","marvell,armada3700-ehci";
+				reg = <0x5e000 0x450>;
+				interrupts = <GIC_SPI 17 IRQ_TYPE_LEVEL_HIGH>;
+				status = "disabled";
+			};
+            utmi_usb32: utmi@5d000 {
+                compatible = "marvell,armada-3700-utmi-phy";
+                reg = <0x5d000 0x1000>;
+               /* utmi-port = <UTMI_PHY_TO_USB3_HOST0>;*/
+                utmi-port = <UTMI_PHY_TO_USB3_DEVICE0>;
+                #phy-cells = <0>;
+                status = "disabled";
+            }; 
+			utmi_usb2: utmi@5f000 {
+				compatible = "marvell,armada-3700-utmi-phy";
+				reg = <0x5f000 0x1000>;
+				utmi-port = <UTMI_PHY_TO_USB2_HOST0>;
+				#phy-cells = <0>;
+				status = "disabled";
+			};
+
+
+			xor@60900 {
+				compatible = "marvell,armada-3700-xor";
+				reg = <0x60900 0x100>,
+				      <0x60b00 0x100>;
+
+				xor10 {
+					interrupts = <GIC_SPI 47 IRQ_TYPE_LEVEL_HIGH>;
+				};
+				xor11 {
+					interrupts = <GIC_SPI 48 IRQ_TYPE_LEVEL_HIGH>;
+				};
+			};
+
+			sdhci0: sdhci@d0000 {
+				compatible = "marvell,armada-3700-sdhci",
+					     "marvell,sdhci-xenon";
+				reg = <0xd0000 0x300>,
+				      <0x1e808 0x4>;
+				interrupts = <GIC_SPI 25 IRQ_TYPE_LEVEL_HIGH>;
+				clocks = <&nb_periph_clk 0>;
+				clock-names = "core";
+				status = "disabled";
+			};
+
+			sdhci1: sdhci@d8000 {
+				compatible = "marvell,armada-3700-sdhci",
+					     "marvell,sdhci-xenon";
+				reg = <0xd8000 0x300>,
+				      <0x17808 0x4>;
+				interrupts = <GIC_SPI 26 IRQ_TYPE_LEVEL_HIGH>;
+				clocks = <&nb_periph_clk 0>;
+				clock-names = "core";
+				status = "disabled";
+			};
+/*
+			sdhci1: sdhci@d0000 {
+				compatible = "marvell,armada-3700-sdhci",
+					     "marvell,sdhci-xenon";
+				reg = <0xd0000 0x300>,
+				      <0x1e808 0x4>;
+				interrupts = <GIC_SPI 25 IRQ_TYPE_LEVEL_HIGH>;
+				clocks = <&sb_periph_clk 10>;
+				clock-names = "core";
+				status = "disabled";
+			};
+
+			sdhci0: sdhci@d8000 {
+				compatible = "marvell,armada-3700-sdhci",
+					     "marvell,sdhci-xenon";
+				reg = <0xd8000 0x300>,
+				      <0x17808 0x4>;
+				interrupts = <GIC_SPI 26 IRQ_TYPE_LEVEL_HIGH>;
+				clocks = <&nb_periph_clk 0>;
+				clock-names = "core";
+				status = "disabled";
+			};
+*/
+			sata: sata@e0000 {
+				compatible = "marvell,armada-3700-ahci";
+				reg = <0xe0000 0x2000>;
+				interrupts = <GIC_SPI 27 IRQ_TYPE_LEVEL_HIGH>;
+				status = "disabled";
+			};
+
+			gic: interrupt-controller@1d00000 {
+				compatible = "arm,gic-v3";
+				#interrupt-cells = <3>;
+				interrupt-controller;
+				reg = <0x1d00000 0x10000>, /* GICD */
+				      <0x1d40000 0x40000>, /* GICR */
+				      <0x1d80000 0x2000>,  /* GICC */
+				      <0x1d90000 0x2000>,  /* GICH */
+				      <0x1da0000 0x20000>; /* GICV */
+				interrupts = <GIC_PPI 9 IRQ_TYPE_LEVEL_HIGH>;
+			};
+		};
+
+		pcie0: pcie@d0070000 {
+			compatible = "marvell,armada-3700-pcie";
+			device_type = "pci";
+			status = "disabled";
+			reg = <0 0xd0070000 0 0x20000>;
+			#address-cells = <3>;
+			#size-cells = <2>;
+			bus-range = <0x00 0xff>;
+		/*	interrupts = <GIC_SPI 29 IRQ_TYPE_LEVEL_HIGH>;*/
+		/*	#interrupt-cells = <1>;*/
+			msi-parent = <&pcie0>;
+			msi-controller;
+			ranges = <0x82000000 0 0xe8000000   0 0xe8000000 0 0x1000000 
+				  0x81000000 0 0xe9000000   0 0xe9000000 0 0x10000>; 
+			/*interrupt-map-mask = <0 0 0 7>;
+			interrupt-map = <0 0 0 1 &pcie_intc 0>,
+					<0 0 0 2 &pcie_intc 1>,
+					<0 0 0 3 &pcie_intc 2>,
+					<0 0 0 4 &pcie_intc 3>;
+			pcie_intc: interrupt-controller {
+				interrupt-controller;
+				#interrupt-cells = <1>;
+			};*/
+		};
+		
+		/*pcie0: pcie@d0070000 {
+			compatible = "marvell,armada-37xx-pcie";
+			reg = <0 0xd0070000 0 0x20000>;
+			#address-cells = <3>;
+			#size-cells = <2>;
+			device_type = "pci";
+			num-lanes = <1>;
+			status = "disabled";
+
+			bus-range = <0 0xff>;
+			ranges = <0x82000000 0 0xe8000000
+				 0 0xe8000000 0 0x1000000
+				 0x81000000 0 0xe9000000
+				 0 0xe9000000 0 0x10000>;
+		};*/
+	};
+};
--- uboot-custom-orig/include/dt-bindings/phy/phy-utmi-mvebu.h	1970-01-01 08:00:00.000000000 +0800
+++ uboot-custom-silicom/include/dt-bindings/phy/phy-utmi-mvebu.h	2020-06-15 11:14:20.933911632 +0800
@@ -0,0 +1,19 @@
+/*
+ * Copyright (C) 2018 Marvell
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ */
+
+#ifndef _DT_BINDINGS_PHY_UTMI_MVEBU
+#define _DT_BINDINGS_PHY_UTMI_MVEBU
+
+#define UTMI_PHY_TO_USB3_HOST0		0
+#define UTMI_PHY_TO_USB3_HOST1		1
+#define UTMI_PHY_TO_USB3_DEVICE0	2
+#define UTMI_PHY_TO_USB2_HOST0		3
+#define UTMI_PHY_INVALID		0xff
+
+#endif /* _DT_BINDINGS_PHY_UTMI_MVEBU */
