diff -uprN arm-trusted-firmware-atf-v1.5-armada-18.12/drivers/marvell/comphy/phy-comphy-3700.c arm-trusted-firmware-atf-v1.5-armada-18.12-silicom/drivers/marvell/comphy/phy-comphy-3700.c
--- arm-trusted-firmware-atf-v1.5-armada-18.12/drivers/marvell/comphy/phy-comphy-3700.c	2018-12-24 22:50:13.000000000 +0800
+++ arm-trusted-firmware-atf-v1.5-armada-18.12-silicom/drivers/marvell/comphy/phy-comphy-3700.c	2020-09-17 13:47:21.232883470 +0800
@@ -549,7 +549,7 @@ static int mvebu_a3700_comphy_usb3_power
 {
 	int ret = 0;
 	uintptr_t reg_base = 0;
-	uint32_t mask, data, addr, cfg, ref_clk;
+	uint32_t mask, data, cfg, ref_clk;
 	void (*usb3_reg_set)(uintptr_t addr, uint32_t reg_offset, uint16_t data,
 			     uint16_t mask, int mode);
 	int mode = COMPHY_GET_MODE(comphy_mode);
@@ -718,12 +718,12 @@ static int mvebu_a3700_comphy_usb3_power
 	udelay(PLL_SET_DELAY_US);
 
 	if (comphy_index == COMPHY_LANE2) {
-		data = COMPHY_LOOPBACK_REG0 + USB3PHY_LANE2_REG_BASE_OFFSET;
+		data = COMPHY_REG_LANE_STATUS1_ADDR + USB3PHY_LANE2_REG_BASE_OFFSET;
 		mmio_write_32(reg_base + COMPHY_LANE2_INDIR_ADDR_OFFSET,
 			      data);
 
-		addr = COMPHY_LOOPBACK_REG0 + USB3PHY_LANE2_REG_BASE_OFFSET;
-		ret = polling_with_timeout(addr, TXDCLK_PCLK_EN, TXDCLK_PCLK_EN,
+		data = reg_base + COMPHY_LANE2_INDIR_DATA_OFFSET;
+		ret = polling_with_timeout(data, TXDCLK_PCLK_EN, TXDCLK_PCLK_EN,
 					   COMPHY_PLL_TIMEOUT, REG_32BIT);
 	} else {
 		ret = polling_with_timeout(LANE_STATUS1_ADDR(USB3) + reg_base,
