--- linux-custom-orig/drivers/tty/tty_ioctl.c	2019-10-21 23:25:45.000000000 +0800
+++ linux-custom-silicom/drivers/tty/tty_ioctl.c	2020-06-23 16:49:01.408926327 +0800
@@ -404,13 +404,14 @@
 			ld->ops->flush_buffer(tty);
 		tty_ldisc_deref(ld);
 	}
-
+	/*Reslove tty can not do hang ioctl(0, SNDCTL_TMR_CONTINUE or TCSETSF*/
+#if 0
 	if (opt & TERMIOS_WAIT) {
 		tty_wait_until_sent(tty, 0);
 		if (signal_pending(current))
 			return -ERESTARTSYS;
 	}
-
+#endif
 	tty_set_termios(tty, &tmp_termios);
 
 	/* FIXME: Arguably if tmp_termios == tty->termios AND the
--- linux-custom-orig/drivers/tty/serial/mvebu-uart.c	2019-10-21 23:25:45.000000000 +0800
+++ linux-custom-silicom/drivers/tty/serial/mvebu-uart.c	2020-06-23 17:14:02.820005870 +0800
@@ -507,8 +507,8 @@
 
 	/* Only the following flag changes are supported */
 	if (old) {
-		termios->c_iflag &= INPCK | IGNPAR;
-		termios->c_iflag |= old->c_iflag & ~(INPCK | IGNPAR);
+		termios->c_iflag &= INPCK | IGNPAR|INLCR |ICRNL;
+		termios->c_iflag |= old->c_iflag & ~(INPCK |IGNPAR| INLCR|ICRNL);
 		termios->c_cflag &= CREAD | CBAUD;
 		termios->c_cflag |= old->c_cflag & ~(CREAD | CBAUD);
 		termios->c_cflag |= CS8;
